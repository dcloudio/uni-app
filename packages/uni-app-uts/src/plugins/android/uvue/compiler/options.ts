import type {
  BindingMetadata,
  CompilerError,
  RootNode,
} from '@vue/compiler-core'
import type { TransformPluginContext } from 'rollup'
import type { RawSourceMap } from 'source-map-js'
import type { DirectiveTransform, NodeTransform } from './transform'

interface SharedTransformCodegenOptions {
  /**
   * @default 'default'
   */
  mode?: 'default' | 'module'
  rootDir?: string
  targetLanguage?: 'kotlin' | 'swift'
  /**
   * Transform expressions like {{ foo }} to `_ctx.foo`.
   * @default false
   */
  prefixIdentifiers?: boolean
  /**
   * Optional binding metadata analyzed from script - used to optimize
   * binding access when `prefixIdentifiers` is enabled.
   */
  bindingMetadata?: BindingMetadata
  /**
   * Compile the function for inlining inside setup().
   * This allows the function to directly access setup() local bindings.
   */
  inline?: boolean
  /**
   * Filename for source map generation.
   * Also used for self-recursive reference in templates
   * @default ''
   */
  filename?: string
  /**
   * 编译的模板类名
   */
  className?: string
  /**
   * 解析 uts component 组件
   * @param name
   * @param type
   */
  parseUTSComponent?: (
    name: string,
    type: 'kotlin' | 'swift'
  ) =>
    | {
        className: string
        namespace: string
        source: string
      }
    | undefined
    | void
  /**
   * 解析 uts customElements 组件
   * @param name
   * @param type
   * @returns
   */
  parseUTSCustomElement?: (
    name: string,
    type: 'kotlin' | 'swift'
  ) =>
    | {
        className: string
        namespace: string
        source: string
      }
    | undefined
    | void
}
export interface CodegenOptions extends SharedTransformCodegenOptions {
  inMap?: RawSourceMap
  /**
   * Generate source map?
   * @default false
   */
  sourceMap?: boolean
  /**
   * 匹配 easycom 组件
   * @param tag
   */
  matchEasyCom?: (
    tag: string,
    uts: boolean
  ) => string | false | undefined | void
  /**
   * template的offset
   */
  originalLineOffset?: number
  /**
   * script的offset
   */
  generatedLineOffset?: number
}

export interface ErrorHandlingOptions {
  onWarn?: (warning: CompilerError) => void
  onError?: (error: CompilerError) => void
}

export interface TransformOptions
  extends SharedTransformCodegenOptions,
    ErrorHandlingOptions {
  rootDir?: string
  /**
   * Cache v-on handlers to avoid creating new inline functions on each render,
   * also avoids the need for dynamically patching the handlers by wrapping it.
   * e.g `@click="foo"` by default is compiled to `{ onClick: foo }`. With this
   * option it's compiled to:
   * ```js
   * { onClick: _cache[0] || (_cache[0] = e => _ctx.foo(e)) }
   * ```
   * - Requires "prefixIdentifiers" to be enabled because it relies on scope
   * analysis to determine if a handler is safe to cache.
   * @default false
   */
  cacheHandlers?: boolean
  /**
   * An array of node transforms to be applied to every AST node.
   */
  nodeTransforms?: NodeTransform[]
  /**
   * An object of { name: transform } to be applied to every directive attribute
   * node found on element nodes.
   */
  directiveTransforms?: Record<string, DirectiveTransform | undefined>
  /**
   * If the pairing runtime provides additional built-in elements, use this to
   * mark them as built-in so the compiler will generate component vnodes
   * for them.
   */
  isBuiltInComponent?: (tag: string) => symbol | void
  /**
   * Used by some transforms that expects only native elements
   */
  isCustomElement?: (tag: string) => boolean | void
  /**
   * SFC scoped styles ID
   */
  scopeId?: string | null
  /**
   * Indicates this SFC template has used :slotted in its styles
   * Defaults to `true` for backwards compatibility - SFC tooling should set it
   * to `false` if no `:slotted` usage is detected in `<style>`
   */
  slotted?: boolean
}

export type TemplateCompilerOptions = {
  /**
   * e.g. platform native elements, e.g. `<div>` for browsers
   */
  isNativeTag?: (tag: string) => boolean

  preprocessLang?: string
  preprocessOptions?: any
  pluginContext?: TransformPluginContext
} & TransformOptions &
  CodegenOptions

export interface CodegenResult {
  ast?: RootNode
  code: string
  preamble?: string
  easyComponentAutoImports: Record<string, [string, string]>
  elements: string[]
  map?: RawSourceMap
}
