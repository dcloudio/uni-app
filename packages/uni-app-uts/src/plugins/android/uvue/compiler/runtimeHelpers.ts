import { once } from '@dcloudio/uni-shared'
import {
  CREATE_COMMENT,
  CREATE_ELEMENT_VNODE,
  CREATE_VNODE,
  NORMALIZE_CLASS,
  NORMALIZE_STYLE,
  TO_DISPLAY_STRING,
  registerRuntimeHelpers,
} from '@vue/compiler-core'
export const IS_TRUE = Symbol(`isTrue`)
export const V_SHOW = Symbol(`vShow`)
export const RENDER_LIST = Symbol(`renderList`)
export const FRAGMENT = Symbol(`Fragment`)
export const OPEN_BLOCK = Symbol(`openBlock`)
export const RESOLVE_COMPONENT = Symbol(`resolveComponent`)
export const RESOLVE_DIRECTIVE = Symbol(`resolveDirective`)
export const RESOLVE_EASY_COMPONENT = Symbol(`resolveEasyComponent`)
export const RENDER_SLOT = Symbol(`renderSlot`)
export const TO_HANDLERS = Symbol(`toHandlers`)

export const V_ON_WITH_MODIFIERS = Symbol(`vOnModifiersGuard`)

export const WITH_SLOT_CTX = Symbol(`withSlotCtx`)
export const WITH_SCOPED_SLOT_CTX = Symbol(`withScopedSlotCtx`)

export const RESOLVE_CACHE = Symbol(`resolveCache`)

export const TRY_SET_REF_VALUE = Symbol(`trySetRefValue`)
export const TRY_UPDATE_REF_NUMBER = Symbol(`tryUpdateRefNumber`)

export const LOOSE_TO_NUMBER = Symbol(`looseToNumber`)

export const initRuntimeHelpersOnce = once(() => {
  registerRuntimeHelpers({
    [IS_TRUE]: 'isTrue',
    [V_SHOW]: 'vShow',
    [RENDER_LIST]: 'RenderHelpers.renderList',
    [FRAGMENT]: 'Fragment',
    [RESOLVE_COMPONENT]: 'resolveComponent',
    [RESOLVE_EASY_COMPONENT]: 'resolveEasyComponent',
    [RESOLVE_DIRECTIVE]: 'resolveDirective',
    [RENDER_SLOT]: `renderSlot`,
    [TO_HANDLERS]: `toHandlers`,
    [V_ON_WITH_MODIFIERS]: `withModifiers`,
    [WITH_SLOT_CTX]: `withSlotCtx`,
    [WITH_SCOPED_SLOT_CTX]: `withScopedSlotCtx`,
    [RESOLVE_CACHE]: `resolveCache`,
    [TRY_SET_REF_VALUE]: `trySetRefValue`,
    [TRY_UPDATE_REF_NUMBER]: `tryUpdateRefNumber`,
    [LOOSE_TO_NUMBER]: `looseToNumber`,
    [CREATE_VNODE]: `_cV`,
    [CREATE_ELEMENT_VNODE]: `_cE`,
    [CREATE_COMMENT]: `_cC`,
    [TO_DISPLAY_STRING]: `_tD`,
    [NORMALIZE_CLASS]: `_nC`,
    [NORMALIZE_STYLE]: `_nS`,
  })
})
