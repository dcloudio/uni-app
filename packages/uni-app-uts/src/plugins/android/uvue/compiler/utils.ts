import path from 'path'
import type { Node } from '@babel/types'
import { type ExpressionNode, createSimpleExpression } from '@vue/compiler-core'
import { walk } from 'estree-walker'
import { parseExpression } from '@babel/parser'
import MagicString from 'magic-string'
import {
  camelize,
  genUTSComponentPublicInstanceIdent,
  normalizePath,
} from '@dcloudio/uni-cli-shared'
import type { TemplateCompilerOptions } from './options'
import { stringifyExpression } from './transforms/transformExpression'
import type { TransformContext } from './transform'
import type { CompilerError } from './errors'

export const __DEV__ = true
export const __BROWSER__ = false
export const __COMPAT__ = false

export function isCompatEnabled(...args: any[]) {
  return false
}

export function genRenderFunctionDecl({
  className,
  genDefaultAs,
  inline = false,
}: TemplateCompilerOptions & { genDefaultAs?: string }): string {
  if (inline) {
    return `(): any | null =>`
  }
  // 调整返回值类型为 any | null, 支持 <template>some text</template>
  const thisCode = genDefaultAs
    ? `this: InstanceType<typeof ${genDefaultAs}>`
    : ''
  return `function ${className || ''}Render(${thisCode}): any | null`
}

export function rewriteObjectExpression(
  exp: ExpressionNode,
  context: TransformContext
) {
  const source = stringifyExpression(exp)
  if (source.includes('{')) {
    const s = new MagicString(source)
    const ast = parseExpression(source, {
      plugins: context.expressionPlugins,
    })
    walk(ast, {
      enter(node: Node) {
        if (node.type === 'ObjectExpression') {
          s.prependLeft(
            node.start!,
            node.properties.length > 0 ? '_uM(' : '_uM<string, any | null>('
          )
          s.prependRight(node.end!, ')')
        }
      },
    })
    return createSimpleExpression(s.toString(), false, exp.loc)
  }
}

export function onCompilerError(error: CompilerError) {}

export function parseSource(fileName: string, rootDir: string) {
  if (fileName.includes('@dcloudio')) {
    return fileName
  }
  rootDir = normalizePath(rootDir)
  if (path.isAbsolute(fileName) && fileName.startsWith(rootDir)) {
    return '@/' + normalizePath(path.relative(rootDir, fileName))
  }
  return fileName
}

export function addEasyComponentAutoImports(
  easyComponentAutoImports: Record<string, [string, string]>,
  rootDir: string,
  tagName: string,
  fileName: string
) {
  if (easyComponentAutoImports[fileName]) {
    return
  }
  // 内置easycom，如 unicloud-db
  if (fileName.includes('@dcloudio')) {
    return
  }
  rootDir = normalizePath(rootDir)
  if (path.isAbsolute(fileName) && fileName.startsWith(rootDir)) {
    fileName = '@/' + normalizePath(path.relative(rootDir, fileName))
  }
  // 加密插件easycom类型导入
  if (fileName.includes('?uts-proxy')) {
    const moduleId = path.basename(fileName.split('?uts-proxy')[0])
    fileName = `uts.sdk.modules.${camelize(moduleId)}`
  }
  const ident = genUTSComponentPublicInstanceIdent(tagName)
  easyComponentAutoImports[fileName] = [ident, ident]
}
