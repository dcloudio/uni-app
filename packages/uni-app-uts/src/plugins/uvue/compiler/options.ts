import { BindingMetadata, CompilerError } from '@vue/compiler-core'
import { RawSourceMap } from 'source-map'
import { DirectiveTransform, NodeTransform } from './transform'

interface SharedTransformCodegenOptions {
  targetLanguage: 'kotlin' | 'swift'
  /**
   * Transform expressions like {{ foo }} to `_ctx.foo`.
   * @default false
   */
  prefixIdentifiers?: boolean
  /**
   * Optional binding metadata analyzed from script - used to optimize
   * binding access when `prefixIdentifiers` is enabled.
   */
  bindingMetadata?: BindingMetadata
  /**
   * Filename for source map generation.
   * Also used for self-recursive reference in templates
   * @default ''
   */
  filename?: string
}
export interface CodegenOptions extends SharedTransformCodegenOptions {
  /**
   * function
   * @default 'default'
   */
  mode?: 'default' | 'function'
  /**
   * Generate source map?
   * @default false
   */
  sourceMap?: boolean
}

export interface ErrorHandlingOptions {
  onWarn?: (warning: CompilerError) => void
  onError?: (error: CompilerError) => void
}

export interface TransformOptions
  extends SharedTransformCodegenOptions,
    ErrorHandlingOptions {
  /**
   * An array of node transforms to be applied to every AST node.
   */
  nodeTransforms?: NodeTransform[]
  /**
   * An object of { name: transform } to be applied to every directive attribute
   * node found on element nodes.
   */
  directiveTransforms?: Record<string, DirectiveTransform | undefined>
  /**
   * If the pairing runtime provides additional built-in elements, use this to
   * mark them as built-in so the compiler will generate component vnodes
   * for them.
   */
  isBuiltInComponent?: (tag: string) => symbol | void
  /**
   * Used by some transforms that expects only native elements
   */
  isCustomElement?: (tag: string) => boolean | void
  /**
   * SFC scoped styles ID
   */
  scopeId?: string | null
  /**
   * Indicates this SFC template has used :slotted in its styles
   * Defaults to `true` for backwards compatibility - SFC tooling should set it
   * to `false` if no `:slotted` usage is detected in `<style>`
   */
  slotted?: boolean
}

export type CompilerOptions = TransformOptions & CodegenOptions

export interface CodegenResult {
  code: string
  map?: RawSourceMap
}
