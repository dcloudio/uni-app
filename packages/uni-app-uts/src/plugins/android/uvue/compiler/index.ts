import { extend } from '@vue/shared'
import { type CompilerError, baseParse } from '@vue/compiler-core'

import { isAppUVueNativeTag } from '@dcloudio/uni-shared'
import {
  createUniVueTransformAssetUrls,
  getBaseNodeTransforms,
  transformTapToClick,
} from '@dcloudio/uni-cli-shared'
import './runtimeHelpers'

import type { CodegenResult, TemplateCompilerOptions } from './options'
import { generate } from './codegen'
import {
  type DirectiveTransform,
  type NodeTransform,
  transform,
} from './transform'
import { transformIf } from './transforms/vIf'
import { transformFor } from './transforms/vFor'
import { transformModel } from './transforms/vModel'
import { transformShow } from './transforms/vShow'
import { transformVText } from './transforms/vText'
import { transformInterpolation } from './transforms/transformInterpolation'
import { transformText } from './transforms/transformText'
import { transformOn } from './transforms/vOnWithModifier'
import { transformBind } from './transforms/vBind'
import { transformSlotOutlet } from './transforms/transformSlotOutlet'
import { transformObjectExpression } from './transforms/transformObjectExpression'
import { transformExpression } from './transforms/transformExpression'
import { transformElements } from './transforms/transformElements'
import { transformStyle } from './transforms/transformStyle'
import { transformVHtml } from './transforms/vHtml'
import { transformMemo } from './transforms/vMemo'
import { transformOnce } from './transforms/vOnce'
import {
  type RawSourceMap,
  SourceMapConsumer,
  SourceMapGenerator,
} from 'source-map-js'
import { trackSlotScopes, trackVForSlotScopes } from './transforms/vSlot'
import { transformElement } from './transforms/transformElement'
import {
  createAssetUrlTransformWithOptions,
  normalizeOptions,
} from '../sfc/compiler/template/transformAssetUrl'
import { createSrcsetTransformWithOptions } from '../sfc/compiler/template/transformSrcset'

export type TransformPreset = [
  NodeTransform[],
  Record<string, DirectiveTransform>
]

export function getBaseTransformPreset(
  prefixIdentifiers?: boolean
): TransformPreset {
  return [
    [
      transformOnce,
      transformIf,
      transformMemo,
      transformFor,
      // order is important
      trackVForSlotScopes,
      transformVHtml,
      transformExpression,
      transformSlotOutlet,
      transformElement,
      trackSlotScopes,
      transformText,
      transformTapToClick,
      transformInterpolation,
      transformObjectExpression,
      transformElements,
      transformStyle,
    ] as any,
    {
      on: transformOn,
      bind: transformBind,
      model: transformModel,
      show: transformShow,
      text: transformVText,
    } as any,
  ]
}

export function compile(
  template: string,
  options: TemplateCompilerOptions
): CodegenResult {
  options.rootDir = options.rootDir || ''
  options.targetLanguage = options.targetLanguage || 'kotlin'
  options.prefixIdentifiers =
    'prefixIdentifiers' in options
      ? options.prefixIdentifiers
      : options.mode === 'module'

  wrapOptionsLog(template, options)

  const isNativeTag =
    options?.isNativeTag ||
    function (tag: string) {
      return (
        isAppUVueNativeTag(tag) ||
        !!options.parseUTSComponent?.(tag, options.targetLanguage!)
      )
    }
  const ast = baseParse(template, {
    comments: false,
    isNativeTag,
    onError: options.onError,
  })
  const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(
    options.prefixIdentifiers
  )

  // 重要不能传入base
  const assetOptions = normalizeOptions(createUniVueTransformAssetUrls(''))

  transform(
    ast,
    extend({}, options, {
      nodeTransforms: [
        ...nodeTransforms,
        createAssetUrlTransformWithOptions(assetOptions),
        createSrcsetTransformWithOptions(assetOptions),
        ...getBaseNodeTransforms('/'),
        ...(options.nodeTransforms || []), // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {} // user transforms
      ),
    })
  )

  const result = generate(ast, options)

  // inMap should be the map produced by ./parse.ts which is a simple line-only
  // mapping. If it is present, we need to adjust the final map and errors to
  // reflect the original line numbers.
  if (options.inMap) {
    if (options.sourceMap) {
      result.map = mapLines(options.inMap, result.map!)
    }
    // if (result.errors.length) {
    //   patchErrors(errors, source, inMap)
    // }
  }

  return result
}

function mapLines(oldMap: RawSourceMap, newMap: RawSourceMap): RawSourceMap {
  if (!oldMap) return newMap
  if (!newMap) return oldMap

  const oldMapConsumer = new SourceMapConsumer(oldMap)
  const newMapConsumer = new SourceMapConsumer(newMap)
  const mergedMapGenerator = new SourceMapGenerator()

  newMapConsumer.eachMapping((m) => {
    if (m.originalLine == null) {
      return
    }

    const origPosInOldMap = oldMapConsumer.originalPositionFor({
      line: m.originalLine,
      column: m.originalColumn ?? 0,
    })

    if (origPosInOldMap.source == null) {
      return
    }

    mergedMapGenerator.addMapping({
      generated: {
        line: m.generatedLine,
        column: m.generatedColumn,
      },
      original: {
        line: origPosInOldMap.line, // map line
        // use current column, since the oldMap produced by @vue/compiler-sfc
        // does not
        column: m.originalColumn ?? 0,
      },
      source: origPosInOldMap.source,
      name: origPosInOldMap.name,
    })
  })

  // source-map's type definition is incomplete
  const generator = mergedMapGenerator as any
  ;(oldMapConsumer as any).sources.forEach((sourceFile: string) => {
    generator._sources.add(sourceFile)
    const sourceContent = oldMapConsumer.sourceContentFor(sourceFile)
    if (sourceContent != null) {
      mergedMapGenerator.setSourceContent(sourceFile, sourceContent)
    }
  })

  generator._sourceRoot = oldMap.sourceRoot
  generator._file = oldMap.file
  return generator.toJSON()
}

function wrapOptionsLog(source: string, options: TemplateCompilerOptions) {
  const { onWarn, onError, inMap } = options
  if (inMap && inMap.sourcesContent?.length) {
    if (onWarn || onError) {
      const originalSource = inMap.sourcesContent![0]
      const offset = originalSource.indexOf(source)
      const lineOffset =
        originalSource.slice(0, offset).split(/\r?\n/).length - 1

      if (onWarn) {
        options.onWarn = (err: CompilerError) => {
          patchError(err, lineOffset, offset)
          onWarn(err)
        }
      }
      if (onError) {
        options.onError = (err: CompilerError) => {
          patchError(err, lineOffset, offset)
          onError(err)
        }
      }
    }
  }
}

function patchError(err: CompilerError, lineOffset: number, offset: number) {
  if (err.loc) {
    err.loc.start.line += lineOffset
    err.loc.start.offset += offset
    if (err.loc.end !== err.loc.start) {
      err.loc.end.line += lineOffset
      err.loc.end.offset += offset
    }
  }
}
