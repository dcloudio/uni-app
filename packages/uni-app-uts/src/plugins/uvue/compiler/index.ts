import { extend } from '@vue/shared'
import {
  baseParse,
  trackSlotScopes,
  trackVForSlotScopes,
  transformElement,
  transformExpression,
} from '@vue/compiler-core'

import { isAppUVueNativeTag } from '@dcloudio/uni-shared'
import { transformTapToClick } from '@dcloudio/uni-cli-shared'
import './runtimeHelpers'

import { CodegenResult, CompilerOptions } from './options'
import { generate } from './codegen'
import { DirectiveTransform, NodeTransform, transform } from './transform'
import { transformIf } from './transforms/vIf'
import { transformFor } from './transforms/vFor'
import { transformModel } from './transforms/vModel'
import { transformShow } from './transforms/vShow'
import { transformVText } from './transforms/vText'
import { transformInterpolation } from './transforms/transformInterpolation'
import { transformText } from './transforms/transformText'
import { transformOn } from './transforms/vOn'
import { transformBind } from './transforms/vBind'
import { transformSlotOutlet } from './transforms/transformSlotOutlet'

export type TransformPreset = [
  NodeTransform[],
  Record<string, DirectiveTransform>
]

export function getBaseTransformPreset(
  prefixIdentifiers?: boolean
): TransformPreset {
  return [
    [
      transformIf,
      transformFor,
      // order is important
      trackVForSlotScopes,
      transformExpression,
      transformSlotOutlet,
      transformElement,
      trackSlotScopes,
      transformText,
      transformTapToClick,
      transformInterpolation,
    ] as any,
    {
      on: transformOn,
      bind: transformBind,
      model: transformModel,
      show: transformShow,
      text: transformVText,
    } as any,
  ]
}

export function compile(
  template: string,
  options: CompilerOptions
): CodegenResult {
  const ast = baseParse(template, {
    isNativeTag(tag) {
      return isAppUVueNativeTag(tag)
    },
  })
  const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(
    options.prefixIdentifiers
  )

  transform(
    ast,
    extend({}, options, {
      prefixIdentifiers: options.prefixIdentifiers,
      nodeTransforms: [
        ...nodeTransforms,
        ...(options.nodeTransforms || []), // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {} // user transforms
      ),
    })
  )

  return generate(ast, options)
}
