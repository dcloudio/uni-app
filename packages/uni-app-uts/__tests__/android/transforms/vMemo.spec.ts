import { compile as baseCompile } from '../../../src/plugins/android/uvue/compiler/index'
describe('compiler: v-memo transform', () => {
  function compile(content: string) {
    return baseCompile(`<view>${content}</view>`, {
      mode: 'module',
      prefixIdentifiers: true,
      filename: 'pages/index/index.uvue',
      className: 'PagesIndexIndex',
    }).code
  }

  test('on root element', () => {
    expect(
      baseCompile(`<view v-memo="[x]"></view>`, {
        mode: 'module',
        prefixIdentifiers: true,
        className: 'PagesIndexIndex',
      }).code
    ).toMatchSnapshot()
  })

  test('on normal element', () => {
    expect(compile(`<view v-memo="[x]"></view>`)).toMatchSnapshot()
  })

  test('on component', () => {
    expect(compile(`<Comp v-memo="[x]"></Comp>`)).toMatchSnapshot()
  })

  test('on v-if', () => {
    expect(
      compile(
        `<view v-if="ok" v-memo="[x]"><text>foo</text>bar</view>
        <Comp v-else v-memo="[x]"></Comp>`
      )
    ).toMatchSnapshot()
  })

  test('on v-for', () => {
    expect(
      compile(
        `<view v-for="{ x, y } in list" :key="x" v-memo="[x, y === z]">
          <text>foobar</text>
        </view>`
      )
    ).toMatchSnapshot()
  })

  test('on v-for with list-item', () => {
    expect(
      compile(
        `<view v-for="{ x, y } in list" :key="x" v-memo="[x, y === z]" list-item>
          <text>foobar</text>
        </view>`
      )
    ).toMatchSnapshot()
  })

  test('on template v-for', () => {
    expect(
      compile(
        `<template v-for="{ x, y } in list" :key="x" v-memo="[x, y === z]">
          <text>foobar</text>
        </template>`
      )
    ).toMatchSnapshot()
  })
})
