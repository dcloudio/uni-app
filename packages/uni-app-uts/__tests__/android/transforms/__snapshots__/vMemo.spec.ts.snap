// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: v-memo transform on component 1`] = `
"function PagesIndexIndexRender(): any | null {
const _ctx = this
const _cache = this.$.renderCache
const _component_Comp = resolveComponent("Comp")

  return createElementVNode("view", null, [
    withMemo([_ctx.x], (): VNode => createVNode(_component_Comp), _cache, 0)
  ])
}"
`;

exports[`compiler: v-memo transform on normal element 1`] = `
"function PagesIndexIndexRender(): any | null {
const _ctx = this
const _cache = this.$.renderCache
  return createElementVNode("view", null, [
    withMemo([_ctx.x], (): VNode => createElementVNode("view"), _cache, 0)
  ])
}"
`;

exports[`compiler: v-memo transform on root element 1`] = `
"function PagesIndexIndexRender(): any | null {
const _ctx = this
const _cache = this.$.renderCache
  return withMemo([_ctx.x], (): VNode => createElementVNode("view"), _cache, 0)
}"
`;

exports[`compiler: v-memo transform on template v-for 1`] = `
"function PagesIndexIndexRender(): any | null {
const _ctx = this
const _cache = this.$.renderCache
  return createElementVNode("view", null, [
    createElementVNode(Fragment, null, RenderHelpers.renderList(_ctx.list, ({ x, y }, __key, __index, _cached): any => {
      const _memo: Array<any | null> = ([x, y === _ctx.z])
      if (_cached != null && _cached.key === x && isMemoSame(_cached, _memo)) return _cached
      const _item = createElementVNode("text", utsMapOf({ key: x }), "foobar")
      _item.memo = _memo
      return _item
    }, _cache, 0), 128 /* KEYED_FRAGMENT */)
  ])
}"
`;

exports[`compiler: v-memo transform on v-for 1`] = `
"function PagesIndexIndexRender(): any | null {
const _ctx = this
const _cache = this.$.renderCache
  return createElementVNode("view", null, [
    createElementVNode(Fragment, null, RenderHelpers.renderList(_ctx.list, ({ x, y }, __key, __index, _cached): any => {
      const _memo: Array<any | null> = ([x, y === _ctx.z])
      if (_cached != null && _cached.key === x && isMemoSame(_cached, _memo)) return _cached
      const _item = createElementVNode("view", utsMapOf({ key: x }), [
        createElementVNode("text", null, "foobar")
      ])
      _item.memo = _memo
      return _item
    }, _cache, 0), 128 /* KEYED_FRAGMENT */)
  ])
}"
`;

exports[`compiler: v-memo transform on v-for with list-item 1`] = `
"function PagesIndexIndexRender(): any | null {
const _ctx = this
const _cache = this.$.renderCache
  return createElementVNode("view", null, [
    createElementVNode(Fragment, null, RenderHelpers.renderList(_ctx.list, ({ x, y }, __key, __index, _cached): any => {
      const _memo: Array<any | null> = ([x, y === _ctx.z])
      if (_cached != null && _cached.key === x && isMemoSame(_cached, _memo)) return _cached
      const _item = createElementVNode("view", utsMapOf({ key: x }), [
        createElementVNode("text", null, "foobar")
      ])
      _item.memo = _memo
      return _item
    }, _cache, 0, "/xVKdEEB476nTdgg8k3dRg=="), 128 /* KEYED_FRAGMENT */)
  ])
}"
`;

exports[`compiler: v-memo transform on v-if 1`] = `
"function PagesIndexIndexRender(): any | null {
const _ctx = this
const _cache = this.$.renderCache
const _component_Comp = resolveComponent("Comp")

  return createElementVNode("view", null, [
    isTrue(_ctx.ok)
      ? withMemo([_ctx.x], (): VNode => createElementVNode("view", utsMapOf({ key: 0 }), [
          createElementVNode("text", null, "foo"),
          "bar"
        ]), _cache, 0)
      : withMemo([_ctx.x], (): VNode => createVNode(_component_Comp, utsMapOf({ key: 1 })), _cache, 1)
  ])
}"
`;
